userspace/libsinsp/threadinfo.cpp

```diff
@@ -505,8 +505,21 @@ void sinsp_threadinfo::set_user(uint32_t uid)
 	scap_userinfo *user = m_inspector->m_usergroup_manager.get_user(m_container_id, uid);
 	if (!user)
 	{
-		// this can fail if import_user is disabled
-		user = m_inspector->m_usergroup_manager.add_user(m_container_id, uid, m_group.gid, NULL, NULL, NULL, m_inspector->is_live());
+		// these can fail if import_user is disabled
+		if(m_container_id.empty())
+		{
+			user = m_inspector->m_usergroup_manager.add_user(m_container_id, uid, m_group.gid, NULL, NULL, NULL, m_inspector->is_live());
+		}
+		else if(uid != 0)
+		{
+			//
+			// When a container is running with a specific user and this
+			// get called with 0, it's too early to make an attempt.
+			// As a downside we won't load users for containers running as
+			// root, but we will load them if e.g.docker exec -u <specific-user>.
+			//
+			user = m_inspector->m_usergroup_manager.add_container_user(m_container_id, m_pid, uid, m_inspector->is_live());
+		}
```

Comment:
Why is it "too early"? Because we caught a thread switching into the container? What if the container _is_ started as root?

What *is* `uid` here? If we're calling this from all sorts of parsers, it's not a special uid, like the `docker -u` user, right? So how does the `uid != 0` check work?

Maybe hardcoding uid=0 as "root" would be good enough btw?

Reply:
@deepskyblue86 :)

Reply:
1. Why is it "too early"?
  By trial and error, I found that the thread had the container id but wasn't really _inside_ the container (i.e. `runc:[1:CHILD]` in my tests).
2. What if the container is started as root?
  We'll have `<NA>` :shrug: but if / when it switches to a different user we'll load the users.
3. What is `uid` here?
  a) `sinsp_threadinfo::set_user` is being called by `parse_chroot_exit`, `parse_clone_exit`, `parse_execve_exit`, `parse_setuid_exit`, `parse_setresuid_exit`.
  b) It's actually the user of `docker -u` or k8s `runAsUser`.
  c) you can see that it's being tested (test/e2e/tests/test_process/test_container.py) by starting a `hashicorp/http-echo:alpine` container with `11:110`
4. Maybe hardcoding uid=0 as "root" would be good enough btw?
  I was a bit uncertain about doing so, but if it's already two of us I'd go with that :grinning:

Reply:
> Maybe hardcoding uid=0 as "root" would be good enough btw?

Makes sense.